### ISSUE TITLE
Retry middleware: custom function to determine when to retry

### ISSUE THREAD

**Issue by venkatd** [teamon/tesla#263]
Currently the middleware pattern matches on `{:ok, env}` or `{:error, reason}`.

https://github.com/teamon/tesla/blob/master/lib/tesla/middleware/retry.ex#L40-L43

However, I believe this logic may be API specific. For example with the HubSpot API, we would want to retry on `{:ok, %{status: 429}}`  (429 = too many requests code) while with other APIs it may depend on what is returned in the body.

I propose adding a `should_retry` option.

**Comment by teamon** [comment]
Good point, I've [updated the docs](https://github.com/teamon/tesla/commit/e03beda86dda1646cb6916c974b1471973d4fb26).

How would your `should_retry` work? Can you share some examples?

**Comment by venkatd** [comment]
@teamon Sure it might looks something like as follows:
```
plug Tesla.Middleware.Retry,
    delay: 10_000,
    max_retries: 5,
    should_retry: fn
      {:ok, %{status: status}} when status in [429, 503] -> true
      {:ok, _} -> false
      {:error, _} -> true
    end
```

See below for the service we are using. Some status codes are worth retrying (like timeouts) while other status codes (unauthorized) are worth moving on with:
https://doc.scrapinghub.com/crawlera.html

**Comment by teamon** [comment]
It makes sense, with the default function of `&match?({:error, _}, &1)`.
@venkatd Would you mind submitting a PR?

**Comment by teamon** [comment]
Closed with #275
